src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+                MaybeUninit::uninit();
+            let mut block: [MaybeUninit<u8>; mem::size_of::<ZipLocalEntryBlock>()] =
+                unsafe { mem::transmute(block) };
+
+            input_file.pread_all(data.header_start, &mut block[..])?;
+
+            let block: MaybeUninit<[u8; mem::size_of::<ZipLocalEntryBlock>()]> =
+                unsafe { mem::transmute(block) };
+            unsafe { block.assume_init() }
+        };
+        // Parse static-sized fields and check the magic value.
+        let block = ZipLocalEntryBlock::interpret(block.as_ref())?;
+
+        // Calculate the end of the local header from the fields we just parsed.
+        let variable_fields_len: u64 =
+            // Each of these fields must be converted to u64 before adding, as the result may
+            // easily overflow a u16.
+            block.file_name_length as u64 + block.extra_field_length as u64;
+        let local_entry_block_size: u64 = mem::size_of::<ZipLocalEntryBlock>().try_into().unwrap();
+        let data_start: u64 = data.header_start + local_entry_block_size + variable_fields_len;
+
+        // Set the value so we don't have to read it again.
+        match data.data_start.set(data_start) {
+            Ok(()) => (),
+            // If the value was already set in the meantime, ensure it matches.
+            Err(_) => {
+                assert_eq!(*data.data_start.get().unwrap(), data_start);
+            }
+        }
+        Ok(data_start)
+    }
+
+    /// Parameters to control the degree of parallelism used for extraction.
+    #[derive(Debug, Clone)]
+    pub struct ExtractionParameters {
+        /// Number of threads used for decompression.
+        ///
+        /// Default value: 4.
+        ///
+        /// Note that multiple times this many threads will be spawned by [`split_extract()`] as
+        /// part of the pipelined process. Only this many threads will be used to perform
+        /// decompression in rust code, but other threads will be used to wait on I/O from
+        /// the kernel.
+        pub decompression_threads: usize,
+        /// Size of buffer used to copy a decompressed entry into the corresponding output pipe.
+        ///
+        /// Default value: 1MB.
+        pub decompression_copy_buffer_length: usize,
+        /// Size of buffer used to copy stored entries into the output file.
+        ///
+        /// Used on non-Linux platforms without
+        /// [`copy_file_range()`](https://www.gnu.org/software/libc/manual/html_node/Copying-File-Data.html),
+        /// as well as on Linux when the input and output file handles are on separate devices.
+        ///
+        /// Default value: 1MB.
+        pub file_range_copy_buffer_length: usize,
+        /// Size of buffer used to splice contents from a pipe into an output file handle.
+        ///
+        /// Used on non-Linux platforms without [`splice()`](https://en.wikipedia.org/wiki/Splice_(system_call)).
```

Comment:
This buffer isn't necessary on _any_ Unix; see https://stackoverflow.com/a/10330172.

Reply:
I'm not sure I understand your meaning here. That answer seems to say that on non-linux platforms, `read()`/`write()` with an explicit buffer (as we do here) is the way to go. Our `PipeReadBufferSplicer` struct performs `read()` then `pwrite_all()` with an explicit buffer, because we can't use `splice()`.

Do I misunderstand you here?